NASA-AMMOS/aerie
# Minimize locks acquired for dataset partitions

* **Tickets addressed:** Related to #595 
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description


**Problem:** Inserting simulation results claims locks that conflict with creating new dataset partitions
**Proposed solution:** Reduce the locks acquired when creating new dataset partitions

Steps taken:
1. Follow [this](https://stackoverflow.com/a/67016755) advice to replace the single-command partition creation with the two-command equivalent. This removes the need to acquire an AccessExclusiveLock on the whole partitioned table when creating a new dataset partition.
2. Replace the following foreign keys with constraint triggers: `profile_segment->profile`, `event->topic`, `span->dataset`. This removes the need to acquire a ShareRowExclusiveLock on `profile`, `topic`, and `dataset` when creating a new dataset partition.
3. Replace the `span->span` self-referencing foreign key with a separate foreign key per partition - e.g. `span_15->span_15`. This removes the need to acquire a ShareRowExclusiveLock on `span` when creating a new dataset partition.

Advantages:
- Inserting simulation results and creating new dataset partitions are independent - neither blocks the other. This allows us to focus on speeding up inserts without worrying about blocking concurrent partition creation.

Disadvantages:
- Triggers add complexity and maintenance burden. Attaching triggers and foreign keys to each partition by hand means that future migrations need to iterate over all partition tables and `alter` them separately.

Alternatives considered:
- Lock no more than one table at a time when inserting.
  This eliminates the deadlock, but attaching a new partition is still blocked until the lock is released - which, for `profile_segment`, can be a very long time
- Guarantee that all inserting transactions commit or rollback within some time limit.
   This is a good practice in general - and I do think we should take steps in this direction - but I do not like the idea of our design hinging on this limit. It introduces tension - the longer the limit, the longer the response time on simulation requests. This effect stacks up if you have alternating inserts and simulation requests. While we can work around this by attaching the partition in a concurrent thread - it becomes quite an intricate dance. At the other extreme, making the limit too short constrains the effort to speed up the writing of simulation results - which is a problem for which I'd like to leave the design space wide open, if we can.

## Verification


To reproduce the deadlock issue on develop:
1. Bring up aerie
9. Stop aerie_merlin_worker_2
10. Attach a debugger to aerie_merlin_worker_1 and set a breakpoint here:
    https://github.com/NASA-AMMOS/aerie/blob/602aeda0456d2fc4514b8966fe46fdc5adf9a49a/merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/ProfileRepository.java#L117
11. Make a plan, run a simulation - observe that the worker stops at the breakpoint
12. Bring up aerie_merlin_worker_2
13. Edit the plan, run another simulation
14. Go back to the debugger for aerie_merlin_worker_1, and hit continue

The postgres log will contain the following:
```
postgres_1                   | 2023-02-06 20:11:49.017 UTC [500] ERROR:  deadlock detected at character 13
postgres_1                   | 2023-02-06 20:11:49.017 UTC [500] DETAIL:  Process 500 waits for RowExclusiveLock on relation 16624 of database 16386; blocked by process 509.
postgres_1                   |  Process 509 waits for ShareRowExclusiveLock on relation 16606 of database 16386; blocked by process 500.
postgres_1                   |  Process 500: insert into profile_segment (dataset_id, profile_id, start_offset, dynamics, is_gap)
postgres_1                   |    values ($1, $2, $3::interval, $4::json, $5)
postgres_1                   |  Process 509: create table profile_segment_63 partition of profile_segment for values in (63)
postgres_1                   | 2023-02-06 20:11:49.017 UTC [500] HINT:  See server log for query details.
postgres_1                   | 2023-02-06 20:11:49.017 UTC [500] STATEMENT:  insert into profile_segment (dataset_id, profile_id, start_offset, dynamics, is_gap)
postgres_1                   |    values ($1, $2, $3::interval, $4::json, $5)
```

To reproduce blocking issue - follow the same steps as above, but stop at this breakpoint instead:
https://github.com/NASA-AMMOS/aerie/blob/602aeda0456d2fc4514b8966fe46fdc5adf9a49a/merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/remotes/postgres/PostgresResultsCellRepository.java#L275

Follow the same steps on this branch, and verify that no deadlock or blocking occurs.

- [x] TODO: Prior to merging, I would like to add db-tests for the new triggers.

## Documentation


- [x] TODO: Prior to merging, add detailed comments to the triggers and prepared statement.

We can mention this in release notes, but I don't think there's a need for user-facing documentation.

I should probably mention this in the SDD though.

## Future work

- When speeding up simulation result writing, we'll likely want to experiment with delaying these triggers, or perhaps rewrite them as per-statement triggers
